File: mobile/lib/features/activities/domain/entities/sales_activity.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import '../../../leads/domain/entities/lead.dart';
import '../../../customers/domain/entities/customer.dart';
import '../../../deals/domain/entities/deal.dart';

part 'sales_activity.freezed.dart';
part 'sales_activity.g.dart';

@freezed
abstract class SalesActivity with _$SalesActivity {
  const SalesActivity._();

  const factory SalesActivity({
    required String id,
    required String title,
    @JsonKey(name: 'user_id') required String userId,
    @JsonKey(name: 'lead_id') String? leadId,
    @JsonKey(name: 'customer_id') String? customerId,
    @JsonKey(name: 'deal_id') String? dealId,
    @JsonKey(name: 'task_destination_id') String? taskDestinationId,
    @JsonKey(name: 'type') required String activityType,
    required String notes,
    @JsonKey(name: 'activity_at') required DateTime activityAt,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
    double? latitude,
    double? longitude,
    @JsonKey(name: 'check_in_time') DateTime? checkInTime,
    @JsonKey(name: 'check_out_time') DateTime? checkOutTime,
    @JsonKey(name: 'selfie_photo_path') String? selfiePhotoPath,
    @JsonKey(name: 'place_photo_path') String? placePhotoPath,
    double? distance,
    @JsonKey(name: 'is_offline') @Default(false) bool isOffline,
    String? address,
    String? outcome,
    Lead? lead,
    Customer? customer,
    Deal? deal,
  }) = _SalesActivity;

  factory SalesActivity.fromJson(Map<String, dynamic> json) =>
      _$SalesActivityFromJson(json);
}

File: backend/internal/domain/models/sales_activity.go
package models

import (
	"crm_wowin_backend/pkg/utils"

	"github.com/google/uuid"
)

type ActivityType string

const (
	ActivityTypeVisit       ActivityType = "visit"
	ActivityTypeNegotiation ActivityType = "negotiation"
	ActivityTypeDealClosing ActivityType = "deal_closing"
	ActivityTypeFollowUp    ActivityType = "follow_up"
	ActivityTypeOther       ActivityType = "other"
)

type SalesActivity struct {
	ID                uuid.UUID      `json:"id"`
	UserID            uuid.UUID      `json:"user_id"`
	LeadID            *uuid.UUID     `json:"lead_id,omitempty"`
	CustomerID        *uuid.UUID     `json:"customer_id,omitempty"`
	DealID            *uuid.UUID     `json:"deal_id,omitempty"`
	DealTitle         *string        `json:"deal_title,omitempty"`
	TaskDestinationID *uuid.UUID     `json:"task_destination_id,omitempty"`
	Type              ActivityType   `json:"type"`
	Title             string         `json:"title"`
	Notes             *string        `json:"notes,omitempty"`
	Latitude          *float64       `json:"latitude,omitempty"`
	Longitude         *float64       `json:"longitude,omitempty"`
	CheckInTime       *utils.FlexTime `json:"check_in_time,omitempty"`
	CheckOutTime      *utils.FlexTime `json:"check_out_time,omitempty"`
	SelfiePhotoPath   *string        `json:"selfie_photo_path,omitempty"`
	PlacePhotoPath    *string        `json:"place_photo_path,omitempty"`
	Distance          *float64       `json:"distance,omitempty"`
	IsOffline         bool           `json:"is_offline"`
	Address           *string        `json:"address,omitempty"`
	Outcome           *string        `json:"outcome,omitempty"`
	SignaturePath     *string        `json:"signature_path,omitempty"`
	CustomerName      *string        `json:"customer_name,omitempty"`
	LeadName          *string        `json:"lead_name,omitempty"`
	DealAmount        *float64       `json:"deal_amount,omitempty"`
	NotaPhotoPath     *string        `json:"nota_photo_path,omitempty"`
	Status            VisitStatus    `json:"status"`
	ActivityAt        utils.FlexTime  `json:"activity_at"`
	CreatedAt         utils.FlexTime  `json:"created_at"`
	UpdatedAt         utils.FlexTime  `json:"updated_at"`
}
